javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/AgentInstaller.java

```diff
@@ -378,5 +380,37 @@ public void onComplete(
     }
   }
 
+  private static class RedefinitionDiscoveryStrategy
+      implements AgentBuilder.RedefinitionStrategy.DiscoveryStrategy {
+    private static final AgentBuilder.RedefinitionStrategy.DiscoveryStrategy delegate =
+        AgentBuilder.RedefinitionStrategy.DiscoveryStrategy.Reiterating.INSTANCE;
+
+    @Override
+    public Iterable<Iterable<Class<?>>> resolve(Instrumentation instrumentation) {
+      // filter out our agent classes and injected helper classes
+      return Iterables.transform(
+          delegate.resolve(instrumentation), i -> Iterables.filter(i, c -> !isIgnored(c)));
+    }
+
+    private static boolean isIgnored(Class<?> c) {
+      ClassLoader cl = c.getClassLoader();
+      if (cl != null && cl.getClass() == AgentClassLoader.class) {
+        return true;
+      }
+
+      if (HelperInjector.isInjectedClass(c)) {
+        return true;
+      }
+
+      if (cl == null) {
+        String name = c.getName();
+        List<String> prefixes = BootstrapPackagePrefixesHolder.getBoostrapPackagePrefixes();
+        return prefixes.stream().anyMatch(prefix -> name.startsWith(prefix));
+      }
```

Comment:
Removing this would remove the question @mateuszrzeszutek raised about `RunnableWrapper`.

Injected classes can't be instrumented since they are injected themselves during `ClassFileTransformer.transform()`, so those are safe to skip, and sounds like are the culprit of the log spamming?

Reply:
Yes, injected classes are the main problem.
I can remove this block but I suspect that it really doesn't exclude anything that should not be excluded. Here is the list of excluded classes from this block from startup -javaagent
```
io.opentelemetry.javaagent.shaded.instrumentation.api.config.Config$$Lambda$55/0x0000000800147440
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.common.ArrayBackedAttributesBuilder$$Lambda$29/0x00000008000ab440
io.opentelemetry.javaagent.shaded.instrumentation.api.config.Config$$Lambda$27/0x00000008000ae440
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.trace.Span
io.opentelemetry.javaagent.shaded.io.opentelemetry.context.ImplicitContextKeyed
io.opentelemetry.javaagent.shaded.io.opentelemetry.context.Context
io.opentelemetry.javaagent.instrumentation.api.concurrent.State
io.opentelemetry.javaagent.instrumentation.api.Pair
io.opentelemetry.javaagent.bootstrap.ExceptionLogger
io.opentelemetry.javaagent.shaded.instrumentation.api.config.MethodsConfigurationParser
io.opentelemetry.javaagent.shaded.instrumentation.api.config.CollectionParsers
io.opentelemetry.javaagent.instrumentation.api.ContextStore$Factory
io.opentelemetry.javaagent.instrumentation.api.ContextStore
io.opentelemetry.javaagent.instrumentation.api.InstrumentationContext
io.opentelemetry.javaagent.bootstrap.FieldBackedContextStoreAppliedMarker
io.opentelemetry.javaagent.instrumentation.api.OpenTelemetrySdkAccess
io.opentelemetry.javaagent.instrumentation.api.OpenTelemetrySdkAccess$ForceFlusher
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.internal.Obfuscated
io.opentelemetry.javaagent.shaded.instrumentation.api.InstrumentationVersion
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.internal.StringUtils
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.common.AutoValue_AttributeKeyImpl
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.common.AttributeKeyImpl
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.common.AttributeType
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.common.AttributeKey
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.internal.Utils
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.common.AutoValue_ArrayBackedAttributes
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.common.ArrayBackedAttributesBuilder
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.common.AttributesBuilder
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.common.ArrayBackedAttributes
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.internal.ImmutableKeyValuePairs
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.common.Attributes
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.metrics.Meter
io.opentelemetry.javaagent.shaded.io.opentelemetry.context.propagation.NoopTextMapPropagator
io.opentelemetry.javaagent.shaded.io.opentelemetry.context.propagation.TextMapPropagator
io.opentelemetry.javaagent.shaded.io.opentelemetry.context.propagation.DefaultContextPropagators
io.opentelemetry.javaagent.shaded.io.opentelemetry.context.propagation.ContextPropagators
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.metrics.MeterProvider
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.trace.TracerProvider
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.DefaultOpenTelemetry$Builder
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.OpenTelemetryBuilder
io.opentelemetry.javaagent.shaded.io.opentelemetry.spi.OpenTelemetryFactory
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.DefaultOpenTelemetry
io.opentelemetry.javaagent.shaded.io.opentelemetry.api.OpenTelemetry
io.opentelemetry.javaagent.shaded.instrumentation.api.config.AutoValue_Config
io.opentelemetry.javaagent.shaded.instrumentation.api.config.Config
io.opentelemetry.javaagent.shaded.instrumentation.api.config.NamingConvention$2
io.opentelemetry.javaagent.shaded.instrumentation.api.config.NamingConvention$1
io.opentelemetry.javaagent.shaded.instrumentation.api.config.NamingConvention
io.opentelemetry.javaagent.shaded.instrumentation.api.config.ConfigBuilder
io.opentelemetry.javaagent.bootstrap.PatchLogger
io.opentelemetry.javaagent.bootstrap.WeakCache
io.opentelemetry.javaagent.bootstrap.WeakCache$Provider
io.opentelemetry.javaagent.instrumentation.api.WeakMap
io.opentelemetry.javaagent.instrumentation.api.WeakMap$Implementation$Default
io.opentelemetry.javaagent.instrumentation.api.WeakMap$Provider
io.opentelemetry.javaagent.instrumentation.api.WeakMap$Implementation
io.opentelemetry.javaagent.shaded.instrumentation.api.internal.BootstrapPackagePrefixesHolder
io.opentelemetry.javaagent.instrumentation.api.SafeServiceLoader
io.opentelemetry.javaagent.bootstrap.InternalJarUrlHandler$InternalJarUrlConnection
io.opentelemetry.javaagent.bootstrap.InternalJarUrlHandler$ResolvedJarEntry
io.opentelemetry.javaagent.bootstrap.InternalJarUrlHandler$FileNotInInternalJar
io.opentelemetry.javaagent.bootstrap.InternalJarUrlHandler
io.opentelemetry.javaagent.bootstrap.AgentClassLoader$BootstrapClassLoaderProxy
io.opentelemetry.javaagent.bootstrap.AgentClassLoader
io.opentelemetry.javaagent.slf4j.impl.OutputChoice$OutputChoiceType
io.opentelemetry.javaagent.slf4j.impl.OutputChoice
io.opentelemetry.javaagent.slf4j.impl.SimpleLoggerConfiguration$1
io.opentelemetry.javaagent.slf4j.impl.SimpleLoggerConfiguration
io.opentelemetry.javaagent.slf4j.impl.SimpleLogger
io.opentelemetry.javaagent.slf4j.helpers.MarkerIgnoringBase
io.opentelemetry.javaagent.slf4j.helpers.NamedLoggerBase
io.opentelemetry.javaagent.slf4j.Logger
io.opentelemetry.javaagent.slf4j.impl.SimpleLoggerFactory
io.opentelemetry.javaagent.slf4j.impl.StaticLoggerBinder
io.opentelemetry.javaagent.slf4j.spi.LoggerFactoryBinder
io.opentelemetry.javaagent.slf4j.helpers.Util
io.opentelemetry.javaagent.slf4j.helpers.NOPLoggerFactory
io.opentelemetry.javaagent.slf4j.helpers.SubstituteLoggerFactory
io.opentelemetry.javaagent.slf4j.ILoggerFactory
io.opentelemetry.javaagent.slf4j.LoggerFactory
io.opentelemetry.javaagent.bootstrap.AgentInitializer
io.opentelemetry.javaagent.shaded.io.opentelemetry.context.propagation.TextMapPropagator$Setter
io.opentelemetry.javaagent.shaded.instrumentation.api.tracer.HttpClientTracer
io.opentelemetry.javaagent.shaded.instrumentation.api.tracer.BaseTracer
```
And from starting a test from ide
```
io.opentelemetry.javaagent.instrumentation.api.concurrent.State
io.opentelemetry.javaagent.bootstrap.ExceptionLogger
io.opentelemetry.javaagent.bootstrap.AgentClassLoader
io.opentelemetry.javaagent.instrumentation.api.ContextStore$Factory
io.opentelemetry.javaagent.instrumentation.api.ContextStore
io.opentelemetry.javaagent.instrumentation.api.InstrumentationContext
io.opentelemetry.javaagent.bootstrap.FieldBackedContextStoreAppliedMarker
io.opentelemetry.javaagent.instrumentation.api.OpenTelemetrySdkAccess
io.opentelemetry.javaagent.instrumentation.api.OpenTelemetrySdkAccess$ForceFlusher
io.opentelemetry.javaagent.bootstrap.AgentClassLoader$BootstrapClassLoaderProxy
io.opentelemetry.javaagent.bootstrap.WeakCache
io.opentelemetry.javaagent.bootstrap.WeakCache$Provider
io.opentelemetry.javaagent.instrumentation.api.WeakMap
io.opentelemetry.javaagent.instrumentation.api.WeakMap$Implementation$Default
io.opentelemetry.javaagent.instrumentation.api.WeakMap$Provider
io.opentelemetry.javaagent.instrumentation.api.WeakMap$Implementation
io.opentelemetry.javaagent.instrumentation.api.SafeServiceLoader
```
If you believe that this block should be removed then let me know.

Reply:
@trask is any action needed here?

Reply:
I think ideally we wouldn't skip any classes during retransform that we don't also skip during normal transform (though I agree it doesn't appear to have any impact currently, it feels like it might lead to a hard-to-find bug down the road).

I'd suggest one of these:
* add the same filter to GlobalIgnoresMatcher, explicitly excluding RunnableWrapper (and see if this is the only weird example like that)
* remove this block for now and we can revisit another time (I think it's potentially a nice startup time optimization not to send these classes to retransform)

Reply:
I removed it.
In GlobalIgnoresMatcher there is
```
    if (name.startsWith("io.opentelemetry.auto.")) {
      // FIXME: We should remove this once
      // https://github.com/raphw/byte-buddy/issues/558 is fixed
      if (name.equals("io.opentelemetry.javaagent.instrumentation.api.concurrent.RunnableWrapper")
          || name.equals(
              "io.opentelemetry.javaagent.instrumentation.api.concurrent.CallableWrapper")) {
        return false;
      }
      return true;
    }
```
I assume "io.opentelemetry.auto." is a remnant from a package rename. Changing auto -> javaagent isn't enough becaue io.opentelemetry.javaagent.instrumentation.awslambda.v1_0.AwsLambdaSqsHandlerTest$TestRequestHandler will be skipped and this cause test failures.

Reply:
oh, nice find! I'll create an issue to track this